lms/services/assignment.py

```diff
@@ -10,55 +15,206 @@ def __init__(self, db):
+    ):
+        assignment = Assignment(
+            tool_consumer_instance_guid=tool_consumer_instance_guid,
+            document_url=document_url,
+            resource_link_id=resource_link_id,
+            ext_lti_assignment_id=ext_lti_assignment_id,
+        )
+        self._db.add(assignment)
+        return assignment
 
     def set_document_url(
-        self, tool_consumer_instance_guid, resource_link_id, document_url
+        self,
+        tool_consumer_instance_guid,
+        document_url,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
     ):
         """
-        Save the given document_url.
+        Save the given document_url in an existing assignment or create new one.
 
         Set the document_url for the assignment that matches
-        tool_consumer_instance_guid and resource_link_id. Any existing document
-        URL for this assignment will be overwritten.
-        """
-        assignment = self.get(tool_consumer_instance_guid, resource_link_id)
+        tool_consumer_instance_guid and resource_link_id/ext_lti_assignment_id
+        or create a new one if none exist on the DB.
 
-        if assignment:
+        Any existing document URL for this assignment will be overwritten.
+        """
+        try:
+            assignment = self.get(
+                tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
+            )
             assignment.document_url = document_url
-        else:
-            self._db.add(
-                Assignment(
-                    document_url=document_url,
-                    resource_link_id=resource_link_id,
-                    tool_consumer_instance_guid=tool_consumer_instance_guid,
-                )
+        except NoResultFound:
+            assignment = self.create(
+                document_url=document_url,
+                resource_link_id=resource_link_id,
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                ext_lti_assignment_id=ext_lti_assignment_id,
             )
 
         # Clear the cache (@lru_cache) on self.get because we've changed the
         # contents of the DB. (Python's @lru_cache doesn't have a way to remove
         # just one key from the cache, you have to clear the entire cache.)
         self.get.cache_clear()
 
+        return assignment
+
+    def _merge_canvas_assignments(self, assignments):
+        # We stored a canvas file assignment before (storing its resource_link_id)
+        # we later configured it (storing its ext_lti_assignment_id) in a new row
+        # and now we are launching it, we want to merge those two assignments
+        old_assignment = (
+            assignments[0] if assignments[0].resource_link_id else assignments[1]
+        )
+        assignment = (
+            assignments[0] if assignments[0].ext_lti_assignment_id else assignments[1]
+        )
+        self._db.delete(old_assignment)
+        self._db.flush()
```

Comment:
Is the flush really necessary here?

Reply:
Yes, otherwise the old and new rows conflict on having the same `resource_link_id` before the delete takes effect 

This could be solved in other ways I think (deferrable constraint? CTEs?) but it all looks more complicated than a flush.

Reply:
Weird. I would think that if you're both deleting an object and updating another object in the same unit of work SQLAlchemy or Postgres would be smart enough to process the deletes first, but I guess not